README.md
# jupyterhub-carina
Carina support for JupyterHub. Authenticate your users with Carina, then run their Jupyter servers on their own Carina clusters.

For an example of what this looks like, head over to [howtowhale.com](https://howtowhale.com) and sign in with your Carina account.

# Installation
`pip install jupyterhub-carina`

# Prerequisites
JupyterHub communicates with Carina using OAuth. Before your JupyterHub installation can integrate with Carina, you must
first register the application. See the [Carina OAuth documentation][carina-oauth] for instructions on how to register. When asked for the **Redirect URI**, use the following format `https://<jupyterhub_domain>/hub/oauth_callback`. The registered callback or redirect URL must match the value in your JupyterHub configuration file.

**Note**: If you have customized the JupyterHub `base_url`, then the Redirect URI must be updated to match.

After you have registered, note the **Application Id**, **Secret** and **Callback URL** because they are required
in order to configure your JupyterHub server to use Carina.

# Configuration

1. Set the `OAUTH_CLIENT_ID` and `OAUTH_CLIENT_SECRET` environment variables, or use the optional configuration values in the next section to specify your OAuth credentials.
2. Update your [JupyterHub's configuration file][jupyterhub-config], **jupyterhub_config.py**, with the required configuration settings. Replace the placeholder values (denoted with angle brackets) with appropriate values.
    * `<carina_username>`: The username or email address that you use to log into [http://getcarina.com][carina].
    * `<jupyterhub_domain>`: The domain name or IP address of your JupyterHub server.

Below is an example configuration file with only the required values specified:

```python
c = get_config()

# Required: Configure JupyterHub to authenticate against Carina
c.JupyterHub.authenticator_class = "jupyterhub_carina.CarinaAuthenticator"
c.CarinaAuthenticator.admin_users = ["<carina_username>"]
c.CarinaAuthenticator.oauth_callback_url = "https://<jupyterhub_domain>/hub/oauth_callback"

# Required: Configure JupyterHub to spawn user servers on Carina
c.JupyterHub.hub_ip = "0.0.0.0"
c.JupyterHub.spawner_class = "jupyterhub_carina.CarinaSpawner"
c.CarinaSpawner.hub_ip_connect = "<jupyterhub_domain>"
```

## Optional Variables
You can customize how the user's Jupyter server is created or how your OAuth credentials are specified.

* `<cluster_name>`: The name of the user's Carina cluster. Defaults to `jupyterhub`.
* `<container_name>`: The name of the Jupyter server container running on the user's cluster. Defaults to `jupyter`.
* `<container_image>`: The name of the image to use for the user's server. Defaults to `jupyter/singleuser`.
*  `<start_timeout>`: The timeout when starting a user's server, this value must account for cluster creation and
    pulling the `container_image`. Defaults to `300` (5 minutes), but may need to be increased depending on the
    size of `container_image`.
* `<cluster_polling_interval>`: The number of seconds between polling for a user's cluster to become active.
    Defaults to `30` seconds.
* `<client_id_env>`: The environment variable containing your Carina OAuth Application Id.
    Defaults to `OAUTH_CLIENT_ID`.
* `<client_secret_env>`: The environment variable containing your Carina OAuth Secret.
    Defaults to `OAUTH_CLIENT_SECRET`.
* `<client_id>`: Your Carina OAuth Application Id. Defaults to the value found in the `<client_id_env>` environment variable.
    For example, you may load this from a config file. _Do NOT hard code this values and check it in!_
* `<client_secret>`: Your Carina OAuth Secret. Defaults to the value found in the `<client_secret_env>` environment variable.
    For example, you may load this from a config file. _Do NOT hard code this values and check it in!_

Below is an example configuration file with all required and optional values specified:

```python
c = get_config()

# Required: Configure JupyterHub to authenticate against Carina
c.JupyterHub.authenticator_class = "jupyterhub_carina.CarinaAuthenticator"
c.CarinaAuthenticator.admin_users = ["<carina_username>"]
c.CarinaAuthenticator.oauth_callback_url = "https://<jupyterhub_domain>/hub/oauth_callback"

# Required: Configure JupyterHub to spawn user servers on Carina
c.JupyterHub.hub_ip = "0.0.0.0"
c.JupyterHub.spawner_class = "jupyterhub_carina.CarinaSpawner"
c.CarinaSpawner.hub_ip_connect = "<jupyterhub_domain>"

# Optional: Tweak how the CarinaSpawner creates the user's Carina cluster and container
c.CarinaSpawner.cluster_name = "<cluster_name>"
c.CarinaSpawner.container_prefix = "<container_prefix>"
c.CarinaSpawner.container_image = "<container_image>"
c.CarinaSpawner.cluster_polling_interval = "<cluster_polling_interval>"

# Optional: Tweak where your Carina OAuth application's credentials are located
c.CarinaAuthenticator.client_id_env = "<client_id_env>"
c.CarinaAuthenticator.client_secret_env = "<client_secret_env>"

# Optional: Directly specify your Carina OAuth application's credentials
c.CarinaAuthenticator.client_id = "<client_id>"
c.CarinaAuthenticator.client_secret = "<client_secret>"
```

[carina]: http://getcarina.com
[carina-oauth]: https://getcarina.com/docs/reference/oauth-integration/#register-your-application
[jupyterhub-config]: http://jupyterhub.readthedocs.org/en/latest/getting-started.html#how-to-configure-jupyterhub


CHANGELOG.md
# Changes in jupyterhub-carina

## 0.1.0
First release

## 0.1.1
* Improve error handling when user cluster is deleted (#4)
* Support JupyterHub/OAuthenticator 0.4 (#2)

## 0.2.0
Support new release of Carina (http://ui.getcarina.com)


jupyterhub_carina/CarinaSpawner.py
import docker
from dockerspawner import DockerSpawner
import os.path
import re
import shutil
from tornado import gen
from traitlets import Dict, Integer, Unicode
from .CarinaOAuthClient import CarinaOAuthClient


class CarinaSpawner(DockerSpawner):
    """
    Spawn the user's Jupyter server on their Carina cluster
    """

    cluster_name = Unicode(
        'jupyterhub',
        help="The name of the user's Carina cluster.",
        config=True)

    # Override the default container name. Since we are running on the user's cluster, we don't
    # need the username suffix
    container_name = Unicode(
        'jupyter',
        help="The name of the Jupyter server container running on the user's cluster.",
        config=True)

    cluster_polling_interval = Integer(
        30,
        help="The number of seconds between polling for a user's cluster to become active.",
        config=True
    )

    # Override the default timeout to allow extra time for creating the cluster and pulling the
    # server image
    start_timeout = Integer(
        300,
        help=DockerSpawner.start_timeout.help,
        config=True)

    # Override the docker run parameters
    extra_host_config = Dict(
        {
            'volumes_from': ['swarm-data'],  # --volumes-from swarm-data
            'port_bindings': {8888: None},  # -p 8888:8888
            'restart_policy': {  # --restart always
                'MaximumRetryCount': 0,
                'Name': 'always'
            },
        },
        help=DockerSpawner.extra_host_config.help,
        config=True)

    def __init__(self, **kwargs):
        # Use a different docker client for each server
        self._client = None
        self._carina_client = None
        self._docker_config = None

        super().__init__(**kwargs)

    @property
    def client(self):
        """
        The Docker client used to connect to the user's Carina cluster
        """
        # TODO: Figure out how to configure this without overriding, or tweak a bit and call super
        if self._client is None:
            creds_dir = self.docker_config['DOCKER_CERT_PATH']
            api_endpoint = self.docker_config['DOCKER_HOST'].replace("tcp://", "https://")
            tls_config = docker.tls.TLSConfig(
                client_cert=(os.path.join(creds_dir, 'cert.pem'),
                             os.path.join(creds_dir, 'key.pem')),
                ca_cert=os.path.join(creds_dir, 'ca.pem'),
                verify=os.path.join(creds_dir, 'ca.pem'),
                assert_hostname=False)

            self._client = docker.Client(version='auto', tls=tls_config, base_url=api_endpoint)

        return self._client

    @property
    def docker_config(self):
        """
        The Docker client configuration variables for the user's Carina cluster
        """
        if self._docker_config is None:
            creds_dir = self.get_user_credentials_dir()
            docker_env = os.path.join(creds_dir, 'docker.env')
            if not os.path.exists(docker_env):
                raise RuntimeError(
                    "ERROR! The credentials for {}/{} could not be found in {}.".format(
                        self.user.name, self.cluster_name, creds_dir))

            with open(docker_env) as f:
                env = f.read()

            host = re.findall("DOCKER_HOST=(.*)", env)[0]
            version = re.findall("DOCKER_VERSION=(.*)", env)[0]

            self._docker_config = {
                'DOCKER_HOST': host,
                'DOCKER_CERT_PATH': creds_dir,
                'DOCKER_TLS_VERIFY': 1,
                'DOCKER_VERSION': version,
            }

        return self._docker_config


    @property
    def carina_client(self):
        if self._carina_client is None:
            self.log.debug("Initializing a carina client for %s", self.user.name)
            # Load OAuth configuration from the authenticator
            cfg = self.authenticator
            self._carina_client = CarinaOAuthClient(cfg.client_id, cfg.client_secret,
                                                    cfg.oauth_callback_url,
                                                    user=self.user.name)

        return self._carina_client

    def get_state(self):
        self.log.debug("Saving state for %s", self.user.name)
        state = super().get_state()
        if self.carina_client.credentials:
            state['access_token'] = self.carina_client.credentials.access_token
            state['refresh_token'] = self.carina_client.credentials.refresh_token
            state['expires_at'] = self.carina_client.credentials.expires_at

        return state

    def load_state(self, state):
        self.log.debug("Loading state for %s", self.user.name)
        super().load_state(state)

        access_token = state.get('access_token', None)
        refresh_token = state.get('refresh_token', None)
        expires_at = state.get('expires_at', None)
        if access_token:
            self.log.debug("Loading users's oauth credentials")
            self.carina_client.load_credentials(access_token, refresh_token, expires_at)

    def clear_state(self):
        self.log.debug("Clearing state")
        super().clear_state()

        # TODO: Move this to DockerSpawner
        self.container_id = ''

    @gen.coroutine
    def get_container(self):
        if not (yield self.cluster_exists()):
            return None

        container = yield super().get_container()
        return container

    def get_env(self):
        env = super().get_env()

        self.log.debug("Adding Docker environment variables to the Jupyter server")
        env['DOCKER_HOST'] = self.docker_config['DOCKER_HOST']
        env['DOCKER_TLS_VERIFY'] = self.docker_config['DOCKER_TLS_VERIFY']
        env['DOCKER_CERT_PATH'] = '/var/run/docker/'
        env['DOCKER_VERSION'] = self.docker_config['DOCKER_VERSION']

        return env

    @gen.coroutine
    def start(self):
        try:
            self.log.info("Creating infrastructure for {}...".format(self.user.name))
            cluster = yield self.create_cluster()
            yield self.download_cluster_credentials(cluster['id'])
            yield self.pull_user_image()

            self.log.info("Starting container for {}...".format(self.user.name))
            yield super().start()

            self.log.debug('Startup for {} is complete!'.format(self.user.name))
        except Exception:
            self.log.exception('Startup for {} failed!'.format(self.user.name))
            raise

    @gen.coroutine
    def create_cluster(self):
        """
        Create a Carina cluster
        Returns existing cluster if one with the same name is already present
        """
        cluster = yield self.carina_client.get_cluster(self.cluster_name)
        if cluster is not None:
            self.log.info("Found existing cluster: {}/{}".format(self.user.name, self.cluster_name))
            return cluster

        self.log.info("Creating cluster {}/{}".format(self.user.name, self.cluster_name))
        return (yield self.carina_client.create_cluster(self.cluster_name))

    @gen.coroutine
    def download_cluster_credentials(self, cluster_id):
        """
        Download the cluster credentials
        """
        credentials_dir = self.get_user_credentials_dir()
        if os.path.exists(credentials_dir):
            return

        self.log.info("Downloading cluster credentials for {}/{} ({})..."
                      .format(self.user.name, self.cluster_name, cluster_id))
        user_dir = "/root/.carina/clusters/{}".format(self.user.name)
        yield self.carina_client.download_cluster_credentials(cluster_id, self.cluster_name,
                                                              user_dir,
                                                              self.cluster_polling_interval)

    @gen.coroutine
    def cluster_exists(self):
        """
        Safely check if the user's cluster exists
        """
        credentials_dir = self.get_user_credentials_dir()
        if not os.path.exists(credentials_dir):
            return False

        try:
            yield self.docker('info')
            return True
        except Exception:
            # Remove old credentials now that they no longer work
            shutil.rmtree(credentials_dir, ignore_errors=True)
            self._client = None
            return False

    @gen.coroutine
    def pull_user_image(self):
        """
        Pull the user image to the cluster
        """
        self.log.debug("Starting to pull {} to the {}/{} cluster..."
                       .format(self.container_image, self.user.name, self.cluster_name))
        yield self.docker("pull", self.container_image)
        self.log.debug("Finished pulling {} to the {}/{} cluster..."
                       .format(self.container_image, self.user.name, self.cluster_name))

    def get_user_credentials_dir(self):
        credentials_dir = "/root/.carina/clusters/{}/{}".format(self.user.name, self.cluster_name)
        return credentials_dir
